riscv_analysis/src/cfg/iterator.rs

```diff
@@ -0,0 +1,88 @@
+use crate::cfg::{Cfg, CfgNode};
+use std::{collections::HashSet, rc::Rc};
+
+/// Iterate over a CFG using the underlying node order.
+///
+/// This iterator is useful when you don't care about the order you iterate over
+/// the nodes.
+pub struct CfgIterator {
+    // We have to store a copy of the nodes, since the user may insert nodes
+    // during the iteration.
+    nodes: Vec<Rc<CfgNode>>,
+    // Where we are in the iteration.
+    index: usize,
+}
+
+impl CfgIterator {
+    /// Create a new iterator over all nodes in the CFG.
+    #[must_use]
+    pub fn new(cfg: &Cfg) -> Self {
+        Self {
+            nodes: cfg.nodes.clone(),
+            index: 0,
+        }
+    }
+}
+
+impl Iterator for CfgIterator {
+    type Item = Rc<CfgNode>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Return the next node, if there is one
+        if let Some(node) = self.nodes.get(self.index) {
+            self.index += 1;
+            return Some(Rc::clone(node));
+        }
+
+        None
+    }
+}
+
+/// Iterate over all CFG nodes reachable from some start node.
+///
+/// This iterator uses the `prevs()` & `nexts()` functions of each CFG node.
```

Comment:
If when you say to call `panic!` you mean when the directions are missing, I think that missing directions are reasonable (For example, a 1 instruction program has no directions). 

One possible solution is to use an option for the `nexts` and `prevs` fields. They could be initialized to `None`, then the direction pass could set them to `Some`. The iterator can then panic if it sees a `None`. This differentiates between uninitialized (`None`) and empty (`Some({})`).

Reply:
Now that I think about it, it might be a lot of work for a little payoff. I think a broader (potential) issue would be enforcing that certain passes were run or produced certain information before others can run. I'll add an issue and mark this as resolved.